nodeup/pkg/model/node_authorizer.go

```diff
@@ -0,0 +1,149 @@
+)
+
+// NodeAuthorizationBuilder is responsible for node authorization
+type NodeAuthorizationBuilder struct {
+	*NodeupModelContext
+}
+
+var _ fi.ModelBuilder = &NodeAuthorizationBuilder{}
+
+// nodeAuthorizationServiceTemplate is the template used by the node client
+var nodeAuthorizationServiceTemplate = `
+Type=oneshot
+EnvironmentFile=/etc/environment
+{{- if .client_file }}
+ExecStartPre=/usr/bin/bash -c 'while [ ! -f {{ .client_cert }} ]; do sleep 5; done; sleep 5'
+{{- end }}
+ExecStartPre=/usr/bin/mkdir -p /var/lib/kubelet
+ExecStart=/usr/bin/docker run --rm --net=host \
+	--volume={{ .kube_config_dir }}:/var/lib/kubelet \
+	--volume={{ .tls_path }}:/config:ro \
+	{{ .image }} \
+	client \
+	--authorizer={{ .authorizer }} \
+	--interval={{ .interval }} \
+	--kubeapi-url={{ .kubeapi_url }} \
+	--kubeconfig={{ .kube_config }} \
+	--node-url={{ .node_url }} \
+	--timeout={{ .timeout }} \
+	--tls-ca=/config/ca.pem \
+	--tls-cert=/config/tls.pem \
+	--tls-private-key=/config/tls-key.pem
+`
+
+// Build is responsible for handling the node authorization client
+func (b *NodeAuthorizationBuilder) Build(c *fi.ModelBuilderContext) error {
+	// @check if we are a master and download the certificates for the node-authozier
+	if b.UseNodeAuthorizer() && b.IsMaster {
+		name := "node-authorizer"
+		// creates /src/kubernetes/node-authorizer/{tls,tls-key}.pem
+		if err := b.BuildCertificatePairTask(c, name, name, "tls"); err != nil {
+			return err
+		}
+		// creates /src/kubernetes/node-authorizer/ca.pem
+		if err := b.BuildCertificateTask(c, fi.CertificateId_CA, filepath.Join(name, "ca.pem")); err != nil {
+			return err
+		}
+	}
+
+	authorizerDir := "node-authorizer"
+
+	// @check if bootstrap tokens are enabled and download client certificates for nodes
+	if b.UseBootstrapTokens() && !b.IsMaster {
+		if err := b.BuildCertificatePairTask(c, "node-authorizer-client", authorizerDir, "tls"); err != nil {
+			return err
+		}
+		if err := b.BuildCertificateTask(c, fi.CertificateId_CA, authorizerDir+"/ca.pem"); err != nil {
+			return err
+		}
+	}
+
+	glog.V(3).Infof("bootstrap: %t, node authorization: %t, node authorizer: %t", b.UseBootstrapTokens(),
+		b.UseNodeAuthorization(), b.UseNodeAuthorizer())
+
+	// @check if the NodeAuthorizer provision the client service for nodes
+	if b.UseNodeAuthorizer() && !b.IsMaster {
+		na := b.Cluster.Spec.NodeAuthorization.NodeAuthorizer
+
+		glog.V(3).Infof("node authorization service is enabled, authorizer: %s", na.Authorizer)
+		glog.V(3).Infof("node authorization url: %s", na.NodeURL)
+
+		// @step: create the systemd unit to run the node authorization client
+		man := &systemd.Manifest{}
+		man.Set("Unit", "Description", "Node Authorization Client")
+		man.Set("Unit", "Documentation", "https://github.com/kubernetes/kops")
+		man.Set("Unit", "After", "docker.service")
+		man.Set("Unit", "Before", "kubelet.service")
+
+		tp, err := template.New("service").Parse(nodeAuthorizationServiceTemplate)
```

Comment:
Oh I see!  I would much rather just keep this as pure code, without the templates - it's much easier to see what's actually happening IMO, as you don't have to jump between the model & the template